File: NumberWarsGamePy/enemy.py
import utils

class Enemy(object):
    def __init__(self, digits = None, possible_guesses = None):
        '''Initializes an instance of the Enemy'''


        if digits is None and possible_guesses is None:
            raise AttributeError("No digits or codes list provided")

        elif digits is not None:
            self._possible_permutations = utils.get_all_codes(digits) # 84.5 secs in 1000 games
        
        elif possible_guesses is not None:
            self._possible_permutations = list.copy(possible_guesses) 

        self._code = utils.get_valid_number(self._possible_permutations)
        self._old_guesses = []
        self._last_guess = None

    @property
    def possible_permutations(self):
        return self._possible_permutations

    @property
    def old_guesses(self):
        return self._old_guesses
    
    @property
    def code(self):
        return self._code

    @property
    def last_guess(self):
        return self._last_guess

    @classmethod
    def enemy_by_given_possibilities(cls, possible_guesses): # 64.7 secs in 1000 games
        '''Initializes an instance of the enemy given the possibilities. This is faster than
        letting the initializer get all the permutations at each object creation'''
        return cls(None, possible_guesses)

    @classmethod
    def enemy_by_digits(cls, digits): # 72.8 secs in 1000 games
        '''Initializes an instance of the enemy given the ammount of digits for 
        the game'''
        return cls(digits, None)
        
    def think(self, good_ammount, regular_ammount):
        '''Compares the last guess to all the remaining possibilities
        and removes any of them that don't match the goods and regulars'''

        for possible_code in list.copy(self.possible_permutations):
            good, regular = utils.answer_to_guess(possible_code, self.last_guess)

            if good_ammount != good or regular_ammount != regular:
                self._possible_permutations.remove(possible_code)

    
    def make_guess(self):
        '''Gets a random number from the remaining possibilities'''

        guess = utils.get_valid_number(self._possible_permutations)

        while guess in self._old_guesses:
            guess = utils.get_valid_number(self._possible_permutations)

        self._last_guess = guess
        self._old_guesses.append(guess)

        return guess

File: NumbersWarGame/Enemy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics;
using System.Threading.Tasks;

namespace NumbersWarGame
{
    class Enemy
    {
        public Level? Difficulty { get; set; }
        public List<string> PossiblePermutations { get; set; }
        public List<string> OldGuesses { get; }
        public string Code { get; }
        internal string LastGuess { get; set; }

        // Common constructor when speed is not an issue
        public Enemy(int Digits)
        {
            PossiblePermutations = new List<string>(Utils.GetAllCodes(Digits));

            OldGuesses = new List<string>();
            Code = Utils.GetValidNumber(PossiblePermutations);
        }

        // Requires the list of permutations to make the class faster and a random generator
        // This constructor is meant for testing
        public Enemy(int Digits, List<string> permutations)
        {
            PossiblePermutations = new List<string>(permutations);

            OldGuesses = new List<string>();
            Code = Utils.GetValidNumber(PossiblePermutations);
        }

        public string MakeGuess()
        {
            string Guess = Utils.GetValidNumber(PossiblePermutations);

            while (OldGuesses.Contains(Guess))
            {
                Guess = Utils.GetValidNumber(PossiblePermutations);
            }

            LastGuess = Guess;
            OldGuesses.Add(Guess);

            return Guess;
        }


        public void Think(string LastGuess, int GoodAmmount, int RegularAmmount )
        {

            // Create a new possible permutations list to be able to modify the original
            foreach (string Possibilities in new List<string>(PossiblePermutations))
            {
                int Good;
                int Regular;
                Utils.AnswerToGuess(LastGuess, Possibilities, out Good, out Regular);

                if (GoodAmmount != Good || RegularAmmount != Regular)
                {
                    PossiblePermutations.Remove(Possibilities);
                }
            }
            return;
        }

        
    }
}
